### DISCUSSION TITLE
Why is the logic for calculating the number of ‘#’ like this?

### DISCUSSION THREAD

**Discussion by Fianax** [pymupdf/pymupdf4llm#201] [Q&A] [answered]
I am creating my own hdr_info_custom for the calculation of the sections.

But I wanted to know how do you do this calculation, i.e. how do you calculate the number of ‘#’?

I've been looking at the code and I got this:

```
fontsizes = {}
for pno in pages:
    page = mydoc.load_page(pno)
    blocks = page.get_text("dict", flags=fitz.TEXTFLAGS_TEXT)["blocks"]
    for span in [  # look at all non-empty horizontal spans
        s
        for b in blocks
        for l in b["lines"]
        for s in l["spans"]
        if not is_white(s["text"])
    ]:
        fontsz = round(span["size"])
        count = fontsizes.get(fontsz, 0) + len(span["text"].strip())
        fontsizes[fontsz] = count
```

Why, in order to count the number of times a word of size X appears, do you count the size of the string, i.e. the number of characters?

Why don't you just count the number of words with that X size?

**Comment by JorjMcKie** [comment]
Word from `get_text("words")` are made from consecutive characters ... of whatever font, font size, etc.
Text extracted by `to_markdown()` extracts spans - which never mix up text particles like this.
Therefore I needed the font size and not the rectangle height of a word.
If you want to fall back to rectangle heights, this roughly probably also works. Use word heights and hope that the `span["bbox"]` is sufficiently close ...

**Reply by Fianax** [reply]
So you used to look at the height of the rectangle containing the word or phrase?

**Reply by Fianax** [reply]
And wouldn't it be better to average the size of each letter in each word to get a more ‘accurate’ ‘size’?

**Reply by JorjMcKie** [reply]
All I'm doing is looking at character bboxes - omitting any white spaces. I am trying to find the bulk of the text "population" in order to set it to the "body text".

The word bbox has a height of the maximum character's height. No way to arrive at the average without extracting single characters = double effort.

**Reply by JorjMcKie** [reply]
And don't forget:
The whole approach has basic issues. E.g. it doesn't look at differences between text styles and or weights. Many headers have the same font size as the following text - but are bold or italic or different color etc.
All that is ignored by my simplifying approach.

Maybe, if the file has a TOC, the "title" items and their hierarchy levels deliver much better approximations. Then the first span of a line would have to be compared to the bookmark items pointing to that page.
Hope I am clear.

**Reply by Fianax** [reply]
And what happens when there are several words of different sizes in the same sentence?

**Reply by JorjMcKie** [reply]
different spans

**Reply by JorjMcKie** [reply, marked as answer]
the first span decides about header level

**Reply by Fianax** [reply]
So in the same blocks[x][‘lines’] there can't be spans with different sizes?

Or in the same blocks[x][‘lines’][y] all `spans` have to have the same size?

```
blocks = page.get_text("dict", flags=pymupdf.TEXTFLAGS_TEXT)["blocks"]
```

**Reply by Fianax** [reply]
but if for example you have this sentence: `hello world HOW ARE YOU`

the lower case part has a font size of 12 and the upper case part has a font size of 16.

You say they are different spans but when creating the sentence, which size is used, 12 or 16?

If size 12 is the predominant size in the document, in this case this sentence will have ‘#’ or not?

In pdf4llm==0.0.7 it would have ‘#’ but in 0.0.9 it would not.

**Reply by JorjMcKie** [reply]
The whole line will be treated as body text because span[0] has body text font size.

**Reply by Fianax** [reply]
So this is possible?

![image](https://github.com/user-attachments/assets/ebf434b7-bb8b-48b3-beff-b7ef431ec2cb)

These two spans are on the same line, so there are two different font sizes on that line. Which one is chosen when this happens?

[prueba_sentence_distinct_size.pdf](https://github.com/user-attachments/files/17979769/prueba_sentence_distinct_size.pdf)

**Reply by JorjMcKie** [reply]
The yellow marked one
![image](https://github.com/user-attachments/assets/93318ab2-28d3-4f06-899f-096dbf450bcd)

**Reply by Fianax** [reply]
then in the same block, line and spans there can be more than one span with different sizes, which is what was shown in the image I showed you.
```
blocks = page.get_text("dict", flags=pymupdf.TEXTFLAGS_TEXT)["blocks"]
```

In this example, there would be two total squares on the same line, right? the `Hey, ` square and the `COMO ESTAS` square?

Because when you create such a line in markdown `Hey, COMO ESTAS` will it have only one size to know whether to put ‘#’ in it or not? which would it be in this case, 12 or 16?

**Reply by JorjMcKie** [reply]
As I wrote above: the code will only pass the first span of a line to hdr_info logic. So whatever nonsense is following after span 0 will not take part in the header level decision.

**Reply by Fianax** [reply]
aaaaaaaaaa

so in this case the line `Hey, COMO ESTAS` will have as size 12 but in this `HEY, como estas` will be 16 because it is only fixed in the first span?

**Reply by JorjMcKie** [reply]
Yes, absolutely. I had hoped to have made that clear several posts before 🤷‍♂️!

**Reply by JorjMcKie** [reply]
Seems I must work on my communication clarity ... 😒

**Reply by Fianax** [reply]
It's my fault for not understanding.

Thank you for explaining it to me

**Reply by JorjMcKie** [reply]
That's what the "Discussions" tab is for ...
